pkg/network/driver_interface.go

```diff
@@ -279,6 +339,43 @@ func resizeDriverBuffer(compareSize int, buffer []uint8) []uint8 {
 	return buffer
 }
 
+// GetDNS returns a raw IP packet that wraps a DNS packet
+func (di *DriverInterface) ReadDNSPacket(visit func([]byte, time.Time) error) (didRead bool, err error) {
+	var bytesRead uint32
+	var key uint32 // returned by GetQueuedCompletionStatus, then ignored
+	var ol *windows.Overlapped
+	err = windows.GetQueuedCompletionStatus(di.dnsIOCP, &bytesRead, &key, &ol, 0) 
+	if err != nil {
+		if err == syscall.Errno(syscall.WAIT_TIMEOUT) {
+			// this indicates that there was no queued completion status, this is fine
+			return false, nil
+		}
+
+		return false, errors.Wrap(err, "could not get queued complation status")
+	}
+
+	var buf *_readbuffer
+	buf = (*_readbuffer)(unsafe.Pointer(ol))
```

Comment:
I don't think this cast is correct. A `*windows.Overlapped` is not equivalent to `*_readbuffer`. You might be able to do this if you change the `_readbuffer` to embed `windows.Overlapped` rather than having it as a field.

Reply:
Another approach would be to have a `map[*windows.Overlapped][4096]byte` to store the association between `windows.Overlapped` structures and the data buffers.